### PR TITLE
Add download support for files

### PR DESCRIPTION
**Pull request by jelly**
Support downloading a single file.

image

Blockers:
 - [x] https://github.com/cockpit-project/cockpit/pull/19890
 - [x] #217 for allowDownload

### CODE REVIEW

**Comment by jelly** [comment]
```
wait_js_cond(ph_in_text(".context-menu li:nth-child(9) button","Delete directory")): Uncaught (in promise) Error: actual text: Download file
```

This is hardcoded and need to be adjusted.

**Comment by jelly** [comment]
Tests do not fail but are flaky.

File: src/common.js
Comment (martinpitt) [inline src/common.js:45]: That doesn't mean that the group has read access.

File: src/common.js
Comment (martinpitt) [inline src/common.js:36]: I'm not super thrilled about this function, but indeed it's cheap and an okay heuristic. Reading may still fail, though.

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:636]: It's a bit arbitrary. There is little reason why this shouldn't work for bigger files, like VM images. This has to be chunked anyway, "all in RAM/single read" files would have to be way smaller than 150 MB, more like 1 MB.

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:647]: Downloading a file through an iframe is something I haven't seen before. How does this handle chunking? There doesn't seem to be any kind of progress information in the code -- is that handled by the browser? Does the fsread1 channel give a proper `Content-Length:` so that the browser knows how much data to expect?

There should be a test case with a huge file (sparse with `truncate` should work fine), to make sure that this doesn't overflow RAM and does chunking correctly.

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:649]: What happens if the download gets interrupted or this fails due to permission errors etc.? I don't see any error handling from the fsread1 channel here.

File: src/sidebar.jsx
Comment (martinpitt) [inline src/sidebar.jsx:349]: Why does this appear both here and in app.jsx? Does this duplicate the menu for some reason?

File: test/check-application
Comment (martinpitt) [inline test/check-application:465]: Please ensure that the downloaded file has exactly this content.

File: test/check-application
Comment (martinpitt) [inline test/check-application:88]: This only checks if it started the downloading, not that it completed. Presumably needs to get passed an expected size or content?

File: test/check-application
Comment (martinpitt) [inline test/check-application:526]: Please add a test for a non-readable file which isn't caught by that heuristic. E.g. the group doesn't have read permission or (if you decide to add that to the heuristic), `chmod u-r`.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:636]: Right, should it be equivalent to the chunk size?

File: src/sidebar.jsx
Comment (jelly) [reply src/sidebar.jsx:349]: Yes, they are different components and duplicated

File: src/common.js
Comment (jelly) [reply src/common.js:36]: I agree, but doing a stat on every file while showing the menu feels rather expensive. I'm open for better options

File: src/sidebar.jsx
Comment (martinpitt) [reply src/sidebar.jsx:349]: Ack.

File: src/common.js
Comment (martinpitt) [reply src/common.js:36]: It's ok, that was more thinking aloud. We need to handle errors, but that's discussed in a separate thread.

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:647]: So we do the same in SoS report to download reports. I'll have to see if we get `Content-length`. I can easily add a test for a huge fle.

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:649]: Right, I guess you just get a "download filed error" from the browser. Setting an alert will be difficult I think.

Comment (martinpitt) [reply src/fileActions.jsx:649]: That's fine -- if this appears as a proper download, then the browser should handle everything.

Comment (martinpitt) [reply src/fileActions.jsx:649]: ... on the browser side. But on the bridge side we probably still need some error handling so that this `Promise` rejects or resolves eventually?

Comment (jelly) [reply src/fileActions.jsx:649]: Check if the load event reports a 403 back or anything not 200 so we can do error handling.

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:639]: let's try and add `content-length:` here.

File: src/fileActions.jsx
Comment (jelly) [inline src/fileActions.jsx:638]: This has a bug! `attachment; filename=\"test.img\"`

File: test/check-application
Comment (jelly) [reply test/check-application:88]: We could try testing cancelDownload as well:
https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-cancelDownload

Not sure how we check if the `fsread` channel is closed however.

For the big file use `truncate -s 1G`

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:639]: That's not allowed to be send over, created this Cockpit issue https://github.com/cockpit-project/cockpit/issues/19861

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:649]: The load event gives us zero information if the url was loaded, neither does onError. I fear we have no way of detection this. One idea is to call cockpit.file.spawn('head -n 4').then(() => { iframe code }).

File: test/check-application
Comment (jelly) [reply test/check-application:465]: Done.

File: test/check-application
Comment (jelly) [reply test/check-application:88]: Added a test to check for expected size, content and one big download.

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:624]: This is now resolved in Cockpit-ws via https://github.com/cockpit-project/cockpit/pull/19889

File: src/app.jsx
Comment (cockpituous) [inline src/app.jsx:191]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/common.js
Comment (cockpituous) [inline src/common.js:50]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/common.js
Comment (cockpituous) [inline src/common.js:53]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:616]: These 10 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:628]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:630]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/sidebar.jsx
Comment (cockpituous) [inline src/sidebar.jsx:348]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-132259-b04878fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/common.js
Comment (cockpituous) [inline src/common.js:50]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-180735-ac57ea47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:628]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-180735-ac57ea47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/sidebar.jsx
Comment (cockpituous) [inline src/sidebar.jsx:348]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240126-180735-ac57ea47-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:616]: I thought setting `content-length` wouldn't work, and instead this is blocked on the cockpit PR?

File: src/fileActions.jsx
Comment (martinpitt) [inline src/fileActions.jsx:612]: I'm not super happy about greedily using superuser perms here. It makes it all too easy to download /etc/shadow or files from a different user. That's a matter of opinion though, and not a veto.

File: test/check-application
Comment (martinpitt) [inline test/check-application:76]: That  happened now. I triggered a lib-update run, so let's land #217 before this.

File: test/check-application
Comment (martinpitt) [inline test/check-application:92]: This shouldn't strip() here, that's weird -- `content` should be exactly as expected

File: test/check-application
Comment (martinpitt) [inline test/check-application:469]: If you don't like the line end here, run `echo -n` or `printf`

File: test/check-application
Comment (martinpitt) [inline test/check-application:479]: Should set expected content.

File: test/check-application
Comment (martinpitt) [inline test/check-application:542]: 500 MiB still fits into the VM's RAM. How long does this take? Could this be 1.5 GiB to be guaranteed to not fit into RAM and thus get read in chunks?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Is this also blocked on https://github.com/cockpit-project/cockpit/pull/19889 ? Well, I suppose we could land this, and download size would automagically work once we land the cockpit fix.

File: test/check-application
Comment (martinpitt) [reply test/check-application:76]: Landed, so please rebase away.

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:616]: Good point, we should drop this. Cockpit-ws should handle this transparently now.

File: test/check-application
Comment (jelly) [reply test/check-application:76]: Thanks!

**Comment by jelly** [comment]
> Thanks! Is this also blocked on [cockpit-project/cockpit#19889](https://github.com/cockpit-project/cockpit/pull/19889) ? Well, I suppose we could land this, and download size would automagically work once we land the cockpit fix.

Not blocked, it is a nice to have show the estimate of the download time but that is totally optional for this PR to land. Also note that navigator is unreleased.

(We also can't test the download estimation I think from CDP)

File: src/common.js
Comment (cockpituous) [inline src/common.js:50]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-094542-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:627]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-094542-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/sidebar.jsx
Comment (cockpituous) [inline src/sidebar.jsx:348]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-094542-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/common.js
Comment (cockpituous) [inline src/common.js:50]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-100727-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx
Comment (cockpituous) [inline src/fileActions.jsx:627]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-100727-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/sidebar.jsx
Comment (cockpituous) [inline src/sidebar.jsx:348]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240129-100727-b86fcbeb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
LGTM now aside from the really dubious `load` handler -- can you please check if this is being called at all, or if this is just a coverage bug?

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:612]: Right, we can drop it because we have checks in place for the file permissions?

**Comment by jelly** [comment]
@martinpitt and then I discovered one "issue", you can't download symlinks:

image


Well I'd say this is a non-bug for now, we need to figure out if the target is a directory or file before we offer downloading. And maybe fsinfo gives us better information for that?

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:627]: I'm pretty sure I tested that the iframe was removed, I had a concern that we would keep creating iframes. But sure let's re-check.


image

Works as expected:

image

File: src/fileActions.jsx
Comment (martinpitt) [reply src/fileActions.jsx:612]: That's the question -- coming from a classical desktop file manager this is weird, and I'd find it nice to at least point out *somehow* (icon, warning, separate "Download as administrator" etc.). But let's not block on this.

File: src/fileActions.jsx
Comment (martinpitt) [reply src/fileActions.jsx:627]: Does the handler get called? I.e. did you add a console.log() to it?

**Comment by martinpitt** [comment]
> And maybe fsinfo gives us better information for that?

Yes,  it very much does. Let's postpone the symlink download thing.

File: src/fileActions.jsx
Comment (jelly) [reply src/fileActions.jsx:627]: Yes, the console log if from the `load event`

**Review by martinpitt — APPROVED** [review]
Dankjewel! (Needs some flake retrying)

**Comment by jelly** [comment]
> Dankjewel! (Needs some flake retrying)

I hope that flake is remedied with `fsinfo` otherwise we need to update the timeout for waiting on `mkdir`

**Comment by jelly** [comment]
@allisonkarlitskaya raised a good point about users who have set ACL's. So maybe we should re-investigate if it's possible to show some sort of cannot download issue. Or just leave it as is and let the user figure it out.

**Comment by allisonkarlitskaya** [comment]
For what it's worth, this is fast and accurate:

```
$ cpf open fsread1 path=/root/x max_read_size=0 | spy -m cockpit.bridge
...
{
  "problem": "access-denied",
  "command": "close",
  "channel": "ch1"
}
```

vs

```
$ cpf open fsread1 path=/etc/passwd max_read_size=0 | spy -m cockpit.bridge
...
{
  "problem": "too-large",
  "command": "close",
  "channel": "ch1"
}
```

**Comment by allisonkarlitskaya** [comment]
I've tried removing the `allowedtoRead()`code and testing this patch:

```diff
diff --git a/src/fileActions.jsx b/src/fileActions.jsx
index 3fb069a..d409a16 100644
--- a/src/fileActions.jsx
+++ b/src/fileActions.jsx
@@ -562,13 +562,12 @@ export const downloadFile = (currentPath, selected) => {
     const prefix = (new URL(cockpit.transport.uri("channel/" + cockpit.transport.csrf_token))).pathname;
     const url = prefix + "?" + query;
     return new Promise((resolve, reject) => {
-        const iframe = document.createElement("iframe");
-        iframe.setAttribute("src", url);
-        iframe.setAttribute("hidden", "hidden");
-        iframe.addEventListener("load", event => {
-            document.body.removeChild(iframe);
-            resolve();
-        });
-        document.body.appendChild(iframe);
+        const a = document.createElement("a");
+        a.href = url;
+        a.download = selected.name;
+        a.hidden = true;
+        document.body.appendChild(a);
+        a.click();
+        document.body.removeChild(a);
     });
 };
```

with somewhat mixed results.

I made my `.bash_logout` non-readable and tried to download it.  Internally -ws is sending the fail.html, complete with a proper error code and message.

Chromium shows that:

image

A quick online search shows no way to control the content of that error message.

Firefox opens an empty tab and shows nothing at all, which feels like a Firefox bug.  I checked that this is also the case for really trivial cases (`python3 -m http.server` on a static html page linking to a missing file) and it does the same thing.  It would be nice if that tab contained the error page...

Cockpit Client doesn't support downloads at all, even when they work :rofl:

File: src/sidebar.jsx
Comment (cockpituous) [inline src/sidebar.jsx:347]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-175-20240131-121935-079ef859-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
This is blocked on Chrom's popup blocking behaviour it blocks files at random without a way for me to figure out why it was blocked. So either we don't test in chromium or we try to fix this behaviour.

**Comment by jelly** [comment]
Closed in favour of https://github.com/cockpit-project/cockpit-files/pull/349

**Pull request closed without merging**
